vyper/old_codegen/parser.py

```diff
@@ -177,8 +178,44 @@ def parse_regular_functions(
     ]
     runtime.extend(internal_funcs)
 
-    # TODO CMC 20210911 why does the lll have a trailing 0
-    o.append(["return", 0, ["lll", runtime, 0]])
+    immutables = [_global for _global in global_ctx._globals.values() if _global.is_immutable]
+
+    if immutables:
+        # find position of the last immutable so we do not overwrite it in memory
+        # when we codecopy the runtime code to memory
+        immutables = sorted(immutables, key=lambda imm: imm._metadata["memory_loc"])
+        start_pos = immutables[-1]._metadata["memory_loc"] + immutables[-1].size * 32
+        # create sequence of actions to copy immutables to the end of the runtime code in memory
+        data_section = []
+        for immutable in immutables:
+            # store each immutable at the end of the runtime code
+            memory_loc, offset = (
+                immutable._metadata["memory_loc"],
+                immutable._metadata["data_offset"],
+            )
+            lhs = LLLnode.from_list(
+                ["add", start_pos + offset, "_lllsz"], typ=immutable.typ, location="memory"
+            )
+            rhs = LLLnode.from_list(memory_loc, typ=immutable.typ, location="memory")
+            data_section.append(make_setter(lhs, rhs, None))
+
+        data_section_size = sum([immutable.size * 32 for immutable in immutables])
+        o.append(
+            [
+                "with",
+                "_lllsz",  # keep size of runtime bytecode in sz var
+                ["lll", runtime, start_pos],  # store runtime code at `start_pos`
+                # sequence of copying immutables, with final action of returning the runtime code
+                ["seq", *data_section, ["return", start_pos, ["add", data_section_size, "_lllsz"]]],
+            ]
+        )
+
+    else:
+        # NOTE: lll macro trailing 0 is the location in memory to store
+        # the compiled bytecode
+        # https://lll-docs.readthedocs.io/en/latest/lll_reference.html#code-lll
+        o.append(["return", 0, ["lll", runtime, 0]])
```

Comment:
Maybe we should reverse the order of the arguments (in vyper/lll/compile_lll.py)